### PR TITLE
Conda recipe

### LINKED ISSUE #349: [REQUEST] Create conda recipe of Open-CE tools
Create a conda recipe of open-CE tools. #329 has added pip package creation mechanism.
We also need scripts to create pip and conda packages of open-ce tools.

### PR DESCRIPTION
**Pull request by npanpaliya**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/open-ce/open-ce/blob/main/CONTRIBUTING.md)?
- [ ] Did you update any affected [documentation](https://github.com/open-ce/open-ce/blob/main/doc/)?
- [ ] Did you write any [tests](https://github.com/open-ce/open-ce/blob/main/tests/) to validate this change?  

## Description

Fixes #349 .

Have created conda recipe along with a basic script to create conda package of open-ce tools.
I think the script can be further enhanced to add a functionality of uploading the package to our channel or may be anaconda's channel (not sure about this), every time we tag the release. I've not done that here. 
Please let me know if we want to handle this in this script (if yes, we will need anaconda's login, etc.. which I'm yet to figure out completely) or it should go to git_tools/create_opence_release.py.

## Review process to land 

1. All tests and other checks must succeed.
2. At least one [maintainer](https://github.com/open-ce/open-ce/blob/main/MAINTAINERS.md) must review and approve.
3. If any  [maintainer](https://github.com/open-ce/open-ce/blob/main/MAINTAINERS.md) requests changes, they must be addressed.

### CODE REVIEW

**Review by bethune-bryant — CHANGES_REQUESTED** [review]
Can we build/install and maybe run a simple test, even something like `open-ce -h`?

Also, does it make sense, (and is it possible) to add a build number check as an optional test to this repo as well, so that we remember to update the build number.

**Comment by npanpaliya** [comment]
@bethune-bryant - I re-read your comment. Please note that I've not created a separate feedstock repo for this. Recipe is pushed  into the same open-ce repo. Do you suggest to create new repo?

**Comment by bethune-bryant** [comment]
> 
> 
> @bethune-bryant - I re-read your comment. Please note that I've not created a separate feedstock repo for this. Recipe is pushed into the same open-ce repo. Do you suggest to create new repo?

No, I don't think we need a separate feedstock, if we can't easily re-use the build number check on this repo we can save that for a future issue.

File: .github/workflows/builder-unit-test.yml
Comment (bethune-bryant) [inline .github/workflows/builder-unit-test.yml:31]: Good catch here!

File: .github/actions/opence-conda-package-required-pr/action.yml
Comment (bethune-bryant) [inline .github/actions/opence-conda-package-required-pr/action.yml:16]: We probably shouldn't make an action for this, since it's only going to be used in this repo. Instead, you can just add this run step directly to the workflow yaml file.

**Review by bethune-bryant — CHANGES_REQUESTED** [review]

File: .github/actions/opence-conda-package-required-pr/action.yml
Comment (npanpaliya) [reply .github/actions/opence-conda-package-required-pr/action.yml:16]: Yeah, earlier I'd added this in the workflow itself. But it was failing to access open-ce root level directory for some reason. I'll retry that change.

Comment (npanpaliya) [reply .github/actions/opence-conda-package-required-pr/action.yml]: Adding it back to workflow and fixing the paths worked. Thanks @bethune-bryant for your review and inputs.

File: .github/workflows/conda-recipe-test.yml
Comment (bethune-bryant) [inline .github/workflows/conda-recipe-test.yml:25]: We should execute open-ce to make sure things are working as a conda package.
```suggestion
          conda list open-ce
          open-ce build env -h

```

File: .github/actions/opence-conda-package-required-pr/action.yml
Comment (bethune-bryant) [reply .github/actions/opence-conda-package-required-pr/action.yml]: Looks good, thanks!

File: .github/workflows/conda-recipe-test.yml
Comment (bethune-bryant) [reply .github/workflows/conda-recipe-test.yml:25]: This is the last thing blocking this PR. I'm going to go ahead and do the commit, and approve the PR so that you can move forward.

**Comment by bethune-bryant** [comment]
I added an `open-ce build env` command to the conda test, and it failed.

Something may not be setup right with the conda package. Is there something extra we need to do for the executable part?

**Comment by npanpaliya** [comment]
> I added an `open-ce build env -h` command to the conda test, and it failed.
> 
> The list worked, but the execution failed:
> 
> ```
> # packages in environment at /usr/share/miniconda:
> #
> # Name                    Version                   Build  Channel
> open-ce                   1.1.3                      py_1    file:///home/runner/work/open-ce/open-ce/open-ce-conda-pkg
> /home/runner/work/_temp/8ad1ed06-63ba-463c-af58-236017df8144.sh: line 5: open-ce: command not found
> ```
> 
> Something may not be setup right with the conda package. Is there something extra we need to do for the executable part?

I've updated the workflow to create a new conda environment to install the conda package of open-ce. And it worked fine. Shall I merge this now?

File: conda/recipe/meta.yaml
Comment (bnemanich) [inline conda/recipe/meta.yaml:12]: I'm not sure we want to explicitly add a build number here. I don't want to have to update the build number every time we make a change to the open-ce repo. Maybe there is something else we can use like the GitHub PR number.

File: conda/recipe/meta.yaml
Comment (npanpaliya) [reply conda/recipe/meta.yaml:23]: conda-build isn't added for pip package too, as we don't expect it to be present when using docker_build.

File: conda/recipe/meta.yaml
Comment (npanpaliya) [reply conda/recipe/meta.yaml:12]: In that case we would even want to remove github workflow tests that should execute `check_build_numbers.py` on this recipe.

File: conda/recipe/meta.yaml
Comment (bnemanich) [reply conda/recipe/meta.yaml:23]: Yeah, but the reasoning for that is we don't want people to have to install conda. In this case, conda is already installed, so installing `conda-build` won't be any more difficult than installing other packages.

Comment (npanpaliya) [reply conda/recipe/meta.yaml:23]: Okay. I'll add at both the places then.

File: conda/recipe/meta.yaml
Comment (npanpaliya) [reply conda/recipe/meta.yaml]: @bnemanich - For build number I've used `GIT_DESCRIBE_NUMBER` which is number of commits since the last tag created on the main branch. So, after every commit/PR merge to main branch it will keep on incrementing until we create a tag. Once the new tag is created, it will be back to 0 and again start incrementing. I think this is the better option than using Github PR number (which can be parsed from $GITHUB_REF) as this variable is empty if we locally try to use it.

There is another side problem that I'm seeing with `git describe --tags` on our open-ce main branch. It shows `open-ce-v1.1.2-38-g6f5c26b` whereas it should have been `open-ce-v1.1.3-35-...`. It could be because 1.1.3 is not tagged from main branch. (_38 is the commit count since 1.1.2 and 35 since 1.1.3 on the main branch_)

File: setup.py
Comment (bethune-bryant) [inline setup.py:29]: Does this requirement check installed conda packages?

Comment (npanpaliya) [reply setup.py:29]: No, with pip, packages listed here would be checked for pip packages instead of conda one. So, if conda-build is already installed via conda, pip might not detect it and install pip package of conda-build too. I've not tried this although.

Comment (bnemanich) [reply setup.py:29]: Yeah, it feels like we shouldn't add this here. I'm not sure what advantage it gives us, and might cause some errors.

Comment (npanpaliya) [reply setup.py]: Removed conda-build.

**Review by bnemanich — APPROVED** [review]

**Review by bethune-bryant — APPROVED** [review]

**Pull request merged**
